### ISSUE TITLE
Defer blocks after return statements are ignored

### ISSUE THREAD

**Issue by rahulghangas** [chapel-lang/chapel#15558]
The example below outputs nothing, while a general interpretation of defer statements in the [docs](https://chapel-lang.org/docs/master/language/spec/statements.html#the-defer-statement) indicates that the defer statement should execute.

$Example$
```
proc foo() {
  return 1;

  defer{
    writeln("Cleanup");
  }
}

foo();
```

$Version$
```chpl version 1.23.0 pre-release (663b1bcf05)```

**Comment by bradcray** [comment]
With the caveat that I'm not an expert on `defer`:

My impression is that the intent of `defer` is that only those `defer` statements that have been reached during execution should execute at the end of the block.  Otherwise, I think they would be far less useful / reliable.  For example, here is what I think of as being a "textbook" use of defer:

```chapel
{
  var myC = new unmanaged C();
  defer {
    delete myC;
  }
}
```

but if we assumed that the defer would run even if the declaration of `myC` was not reached:

```chapel
{
  if exitEarly then return;
  var myC = new unmanaged C();
  defer {
    delete myC;
  }
}
```

then we've made the program incorrect by deleting something that was never allocated or initialized.

For that reason, my impression is that it's probably the docs rather than the implementation that need updating, perhaps by tightening up the phrase `the in-scope defer statements"...? in:

> At a given place where control flow exits a block, the in-scope defer statements and the local variables will be handled in reverse declaration order. Handling a defer statement consists of executing the contained clean-up action. Handling a local variable consists of running its deinitializer if it is of record type.

**Comment by ram-nad** [comment]
go also defers only those statements that have been reached during execution.

**Comment by mppf** [comment]
I agree with @bradcray - the next step here is to adjust the wording of the documentation.

**Comment by ram-nad** [comment]
Can I take these up?

**Comment by mppf** [comment]
Sure, you can look at adjusting the wording of the documentation.
